solutions/0300-0399/0312-burst-balloons-hard.md

```diff
@@ -0,0 +1,163 @@
+---
+description: 'Author: @AnshikaAnand222 | https://leetcode.com/problems/burst-balloons/'
+tags: [Dynamic Programming, Recursion]
+---
+
+# 0312 - Burst Balloons - (Hard) 
+
+## Problem Link
+
+https://leetcode.com/problems/burst-balloons/
+
+## Problem Statement
+
+You are given `n` balloons, indexed from `0` to `n - 1`. Each balloon is painted with a number on it represented by an array `nums`. You are asked to burst all the balloons.
+If you burst the ith balloon, you will get `nums[i - 1] * nums[i] * nums[i + 1]` coins. If `i - 1` or `i + 1` goes out of bounds of the array, then treat it as if there is a balloon with a 1 painted on it.
+
+*Return the maximum coins you can collect by bursting the balloons wisely.*
+
+
+**Example 1:**
+
+```
+Input: nums = [3,1,5,8]
+Output: 167
+
+Explanation:
+nums = [3,1,5,8] --> [3,5,8] --> [3,8] --> [8] --> []
+coins =  3*1*5    +   3*5*8   +  1*3*8  + 1*8*1 = 167
+
+```
+
+**Example 2:**
+
+```
+Input: nums = [1,5]
+Output: 10
+```
+
+
+**Constraints:**
+
+```
+n == nums.length
+1 <= n <= 300
+0 <= nums[i] <= 100
+```
+
+
+## Approach 1: Dynamic Programming
+
+**Analysis**
+
+We need to find a way to divide the problems. If we start from the first balloon, we can't determine the left/right for the number in each sub-problem, If we start from the last balloon, we can.
+We can see the transformation equation is very similar to the one for matrix multiplication.
+
+`dp[i][j] = max(dp[i][j], nums[i] * nums[k] * nums[j] + dp[i][k] + dp[k][j]) # i < k < j`
+
+This is a typical interval DP problem. Because the order of the number extracted matters, we need to do a O(n^3) DP. If we only need to expand the interval to the left or right, we only need to do a O(n^2) DP.
```

Comment:
1. apply LaTex on formula & big o notation
2. need more explanations for beginners

Reply:
1. LaTex ....

```
dp[i][j] = max(dp[i][j], nums[i] * nums[k] * nums[j] + dp[i][k] + dp[k][j]) # i < k < j
```

2. Need explanation on the DP transition

Reply:
I already added LaTex on formula